fairscale/nn/data_parallel/sharded_ddp.py

```diff
@@ -205,13 +237,35 @@ def to(  # type: ignore
 
         """
 
-        for optimizer in self.buckets.keys():
-            for device in self.buckets[optimizer].keys():
-                for bucket in self.buckets[optimizer][device]:
-                    bucket.buffer.to(device=device, dtype=dtype, non_blocking=non_blocking)
+        for device in self.buckets.keys():
+            for bucket in self.buckets[device]:
+                bucket.buffer.to(device=device, dtype=dtype, non_blocking=non_blocking)
 
         self.module.to(device)
 
+    def refresh_trainable(self) -> None:
+        """ If the module trainability has changed, update all the assumptions """
+
+        self._trainable_params = list(filter(lambda x: x.requires_grad, self._all_params))
```

Comment:
anything to assert (assumptions) up on entering this function?

Reply:
good point, I'll try to think about something, all this is tricky (to me at least) so I could do with more asserts indeed

Reply:
definitely not "just to you"! 🤣

Reply:
pulling my hairs now on issues with the state dict loading & custom optimizers (turns out I should not write in their state), distributed training can certainly get a tad bit complex...

Reply:
feel you!

Reply:
I just added an assert, it was a very good call I think, for instance somebody could have had the 'no_sync' context activated, then refresh trainability while forgetting to send the gradients (which would be lost).